fireant/slicer/queries/builder.py

```diff
@@ -0,0 +1,237 @@
+from collections import defaultdict
+
+from fireant.utils import (
+    immutable,
+    ordered_distinct_list,
+    ordered_distinct_list_by_attr,
+)
+from toposort import (
+    CircularDependencyError,
+    toposort_flatten,
+)
+
+from .database import fetch_data
+from .references import join_reference
+from ..exceptions import (
+    CircularJoinsException,
+    MissingTableJoinException,
+)
+from ..filters import DimensionFilter
+from ..intervals import DatetimeInterval
+from ..operations import Operation
+
+
+def _build_dimension_definition(dimension, interval_func):
+    if hasattr(dimension, 'interval') and isinstance(dimension.interval, DatetimeInterval):
+        return interval_func(dimension.definition,
+                             dimension.interval).as_(dimension.key)
+
+    return dimension.definition.as_(dimension.key)
+
+
+class QueryBuilder(object):
+    """
+
+    """
+
+    def __init__(self, slicer):
+        self.slicer = slicer
+        self._widgets = []
+        self._dimensions = []
+        self._filters = []
+        self._orders = []
+
+    @immutable
+    def widget(self, *widgets):
+        """
+
+        :param widget:
+        :return:
+        """
+        self._widgets += widgets
+
+    @immutable
+    def dimension(self, *dimensions):
+        """
+
+        :param dimension:
+        :return:
+        """
+        self._dimensions += dimensions
+
+    @immutable
+    def filter(self, *filters):
+        """
+        :param filter:
+        :return:
+        """
+        self._filters += filters
+
+    @property
+    def tables(self):
+        """
+        :return:
+            A collection of tables required to execute a query,
+        """
+        return ordered_distinct_list([table
```

Comment:
That's a very convoluted list comprehension. Breaking it down could improve clarity in my and @mikeengland opinion.